### PR TITLE
JAX version of DQN

### PR DESCRIPTION
**Pull request by FangLinHe**
# Why to use JAX instead of PyTorch?

1. **Relatively lightweight**: I have noticed that PyTorch is now too powerful for the question in hand. All I need to use is to build simple network (few linear layers, typical activation layers, etc.), optimize weights with simple loss functions (Huber loss, L2 loss, etc.). For example, the [latest build of PyTorch v2.3.0](https://github.com/pytorch/pytorch/releases/tag/v2.3.0) has the file size (265 MB), but [JAX v0.4.28](https://github.com/google/jax/releases/tag/jax-v0.4.28) is only 14.1 MB. I also used [Flax v0.8.3](https://github.com/google/flax/releases/tag/v0.8.3) to build networks and use its optimizer, which is also very light weight, only 2.9 MB.
2. **Just to learn and practice using new tools**: I have learned JAX last year when I did my capstone project at ETHZ, but eventually I didn't need to write much code with JAX. Still, being curious to see if I can replace PyTorch with it, so I decided to use this project as a good starting point. Btw, kudos ⭐ to my supervisor Simon and his [JAX-based simulation-based inference library](https://github.com/dirmeier/sbijax). I definitely took a lot of inspirations from this repository.

# What is my experience using JAX?

## Code

In overall, I might be just too familiar with PyTorch, so I have noticed quite many differences in how the code is written. For example, for a simple network:

```
# PyTorch
import torch
from torch import Tensor, nn, optim


class Network(nn.Module):
    def __init__(self):
        nn.Module.__init__(self)
        self.linear1: nn.Module = nn.Linear(3, 5)
        self.relu = nn.ReLU()
        self.linear2: nn.Module = nn.Linear(5, 2)

    def forward(self, x):
        x = self.linear1(x)
        x = self.relu(x)
        return self.linear2(x)


network = Network()
optimizer = optim.Adam(network.parameters())

input_tensor: Tensor = torch.tensor(
    [[-1, -2, -3], [1, 2, 3]], dtype=torch.float32
)  # Batch size 2, input size 3
output_tensor: Tensor = network(input_tensor)
target_tensor: Tensor = torch.tensor([[1, 2], [3, 4]], dtype=torch.float32)

loss_fn = nn.HuberLoss()
loss = loss_fn(output_tensor, target_tensor)

optimizer.zero_grad()
loss.backward()
optimizer.step()
```

```
# JAX / Flax
from typing import Any, Tuple
import jax
import jax.numpy as jnp
from flax import linen as nn
from flax.training.train_state import TrainState
import optax


class Network(nn.Module):
    @nn.compact
    def __call__(self, x: jax.Array) -> jax.Array:
        x = nn.Dense(features=5)(x)
        x = nn.relu(x)
        return nn.Dense(features=2)(x)


def create_train_state(rng: jax.Array, net: nn.Module) -> TrainState:
    variables = net.init(rng, jnp.ones((1, 3), dtype=jnp.float32))  # input size 3
    tx: optax.GradientTransformation = optax.adam(0.001)
    return TrainState.create(
        apply_fn=net.apply,
        params=variables["params"],
        tx=tx,
    )


@jax.jit
def train_step(
    train_state: TrainState, inputs: jax.Array, targets: jax.Array
) -> Tuple[TrainState, Any]:
    def loss_fn(params) -> jax.Array:
        pred = train_state.apply_fn(
            {"params": params},
            inputs,
        )
        loss = jnp.mean(optax.huber_loss(pred, targets))

        return loss

    grad_fn = jax.value_and_grad(loss_fn, has_aux=False)
    loss, grads = grad_fn(train_state.params)

    train_state = train_state.apply_gradients(grads=grads)

    return train_state, loss


rng: jax.Array = jax.random.key(0)
network = Network()
train_state = create_train_state(rng, network)

inputs: jax.Array = jnp.array(
    [[-1.0, -2.0, -3.0], [1.0, 2.0, 3.0]], dtype=jnp.float32
)  # Batch size 2, input size 3
targets: jax.Array = jnp.array([[1.0, 2.0], [3.0, 4.0]], dtype=jnp.float32)
train_state, loss = train_step(train_state, inputs, targets)
print(loss)
```

We can see that the network construction is very similar, but how to train the model is much more complex, for example, parameters must be created separately and passed around; there is "train state" which includes apply function from the network, state during the training, and even the network weights. But then we still need to pass weights around when we compute the loss or make a forward pass. Even worse, after getting the gradients, we must apply it to the training state and manually update the state in the program.

In overall, that is fine if it simply works. What annoyed me the most is the less useful debug message and lack of actual line information. Every time it took me more than few hours to figure out a very small bug, for example, writing `nn.relu()(x)` with the redundant parenthesis, or passing only `params` to `train_state.apply_fn` without the dictionary.

Maybe it's also because I didn't read the document from the beginning very well, so I didn't understand the concept well enough to figure out the errors.

### Where the weights are stored

In PyTorch, weights gradients are stored in `nn.Module` itself, so we only need to keep the network in one place. But in JAX / Flax, training states, weights, gradients, losses, etc. are all returned and need to be managed by ourselves, so the readability is not that great.

## Performance on CPU

When I finished the first working version, I realized that the total training time is almost twice compared to the PyTorch implementation, which was very disappointing. I don't really want to sacrifice the performance for package size. Then after some effort, I found that making jnp.array from lists of lists is very slow. From [this discussion](https://github.com/google/jax/issues/10662), I then found out I could simply make it more than 10x faster by converting to `np.array` first, then `jnp.arary`. To be honest, I don't like it.

## Documentation

I found out that the documentation is not well organized. And I remembered some of the tutorials were not up-to-date. Many functions do not provide simple examples in their docstring.

But sometimes they appear to be very helpful, for example, [BatchNorm](https://flax.readthedocs.io/en/latest/guides/training_techniques/batch_norm.html) layers require manually changing "few" lines of code in order to make it work, as it contains `batch_stats` that we need to pass around and update. Following the document made the work slightly easier. Only a bit.

In general, I still haven't found any benefits of using JAX over PyTorch, maybe I will figure it out later if I use it more. But so far, it still has lots of rooms for improvement, and before it gets better, I may not try again in the future for simple projects.

### CODE REVIEW

**Comment by FangLinHe** [comment]
Still missing:
1. Run evaluation with JAX model
